no-op pefile==2024.8.26's use of gc.collect() (#8762)

bwoodsend opened the pull request:
The performance regression caused by pefile's explicit garbage collect doesn't look like it's going away any time soon (if ever). As an alternative to continuing to block users from upgrading their pefile version, replace pefile's reference to gc.collect() with a placebo function to *disable* the garbage collect.

bwoodsend committed:
no-op pefile==2024.8.26's use of gc.collect() (#8762)

The performance regression caused by pefile's explicit garbage collect
doesn't look like it's going away any time soon (if ever). As an
alternative to continuing to block users from upgrading their pefile
version, replace pefile's reference to gc.collect() with a placebo
function to *disable* the garbage collect.

bwoodsend commented:
Not sure if I really feel any better about doing this than I do about the pefile version constraint. What do you think?

rokm commented:
> Not sure if I really feel any better about doing this than I do about the pefile version constraint. What do you think?

I was leaning towards vendoring our copy of `pefile` with a fix, but I like the idea of making `gc.collect()` no-op within the context of our `pefile` use better.

rokm commented:
On a semi-related note - how about if  instead of vendoring `altgraph` (which also seems to have little chance of seeing an update) to address the impending disappearance of `pkg_resources`, we also provided a similar fallback as part of our vendored `modulegraph`?

bwoodsend committed:
Prevent altgraph from importing pkg_resources

pkg_resources is scheduled to be removed in 2025-11-30 but altgraph
still uses it to query its own version. There is an open pull request to
altgraph to fix it (ronaldoussoren/altgraph#25) but it's showing no
signs getting merged.

Monkeypatch a fake pkg_resources into sys.modules when importing
altgraph so that its version query doesn't take out PyInstaller come
2025-11-30.

bwoodsend commented:
I have been trying to get rid of the altgraph dependency, either by editing our current modulegraph code or by trying to get #9159 to even build. Both are feeling a bit impossible...

bwoodsend commented:
Oops, didn't spot that you'd done the altgraph one too.

rokm commented on PyInstaller/lib/modulegraph/modulegraph.py:
```diff
@@ -34,14 +35,19 @@
-        category=UserWarning,
-        message="pkg_resources is deprecated",
-    )
+try:
```
The above comment should probably be updated (or removed altogether).

rokm commented:
Hmm, [those failures in Cygwin workflow](https://github.com/pyinstaller/pyinstaller/actions/runs/17883473687/job/50854120713) actually look legit for once...

rokm committed:
hooks: add pre-safe-import-module hooks for `backports` and `jaraco`

Add pre-safe-import-module hooks for `backports` and `jaraco`, in
order to properly handle `setuptools`-vendored copies of these
top-level namespace (or namespace-like) packages in situations
when the `setuptools` vendored-package directory is not in the
python's search path at the start of analysis process.

Both packages have namespace-package semantics, and can therefore
be split between multiple locations (i.e., mixture of components
in regular site-packages and vendored components in `setuptools`
vendor directory); this requires additional considerations compared
to other vendored packages and modules (for which we have already
set up hooks). If such a top-level namespace package is *fully
vendored* (i.e., is provided solely by `setuptools`), we need to
set up an alias (which prevents creation of unnecessary top-level
package in PYZ archive). However, if it is *partially vendored*,
we need to extend the search paths for the package, to accommodate
corner case where `setuptools` vendored-package directory is not
in the search path at the start of analysis process.

rokm commented:
Turns out that the import of `pkg_resources` in `altgraph` was implicitly guaranteeing that at the start of analysis, we always had `setuptools` imported and thus its vendor directory in `sys.path`.

Which in turn masked the fact that we do not have pre-safe-import-module hooks for `backports` and `jaraco`, only for their vendored subpackages/submodules. So if analysis happens to be ran without `setuptools` vendor directory in `sys.path`, these vendored subpackages/submodules fail to be discovered because the top-level namespace (or namespace-like, in case of `backports`) package itself cannot be discovered.

I'm not particularly happy with the fact that none of other CI workflows caught this issue, but I suppose that once `setuptools` happens to be imported as part of one test, that whole worker is "tainted" w.r.t. this issue...

bwoodsend commented:
> I'm not particularly happy with the fact that none of other CI workflows caught this issue, but I suppose that once setuptools happens to be imported as part of one test, that whole worker is "tainted" w.r.t. this issue...

Did you find where this happens? I can't find any imports to setuptools or pkg_resouces that aren't safely tucked away behind `@isolated.decorate`. I also couldn't flush out anything by dotting some `sys.modules["setutools"] = None` about the code base.

rokm commented:
> > I'm not particularly happy with the fact that none of other CI workflows caught this issue, but I suppose that once setuptools happens to be imported as part of one test, that whole worker is "tainted" w.r.t. this issue...
> 
> Did you find where this happens? I can't find any imports to setuptools or pkg_resouces that aren't safely tucked away behind `@isolated.decorate`. I also couldn't flush out anything by dotting some `sys.modules["setutools"] = None` about the code base.

I assumed it happens during modulegraph analysis - i.e., when we are trying to look-up some sub-module of `setuptools` in the modulegraph, the top-level package needs to be imported?

bwoodsend commented:
Yeah, that would explain it. And it makes me feel a lot less confident in all our process isolation stuff...
